inc/saihostif.h

```diff
@@ -343,6 +343,13 @@ typedef enum _sai_hostif_trap_type_t
      */
     SAI_HOSTIF_TRAP_TYPE_ISIS = 0x00002014,
 
+    /**
+     * @brief Packets matching subnet routes with NH pointing to router interface
+     * i.e., no neighbor entry route is present
+     * (default packet action is trap)
+     */
+    SAI_HOSTIF_TRAP_TYPE_NEIGHBOR_MISS = 0x00002015,
+
```

Comment:
Why we can not use the following user define trap for subnet miss ?
    
/**
     * @brief Neighbor table traps
     *
     * Generated by neighbor table entry hit with action trap/log, or by neighbor table miss
     */
    SAI_HOSTIF_USER_DEFINED_TRAP_TYPE_NEIGHBOR,

Reply:
Marian had raised this issue during the community meeting. Repeating my explanation here: 

SAI Hostif User Defined Trap (UDT) behavior is different from default SAI Hostif traps.

```c
/**
* @brief Host interface user defined trap type
*
* User defined traps action is controlled by the referencing object.
* For example, ACL entry with packet action trap and user trap object ID
*/
```

UDT objects have to be associated with corresponding SAI objects. For e.g., SAI\_HOSTIF\_USER\_DEFINED\_TRAP\_TYPE\_ACL are associated with ACL entries having a packet action like trap, copy, log etc. In the case we are talking about, there is no Neighbor entry object in the first place to associate this UDT.

 Creating a default hostif trap allows the NOS to have the same workflow as they do for existing default traps.

Reply:
Marian had raised this issue during the community meeting. Repeating my explanation here: 

SAI Hostif User Defined Trap (UDT) behavior is different from default SAI Hostif traps.

```c
/**
* @brief Host interface user defined trap type
*
* User defined traps action is controlled by the referencing object.
* For example, ACL entry with packet action trap and user trap object ID
*/
```

UDT objects have to be associated with corresponding SAI objects. For e.g., SAI\_HOSTIF\_USER\_DEFINED\_TRAP\_TYPE\_ACL are associated with ACL entries having a packet action like trap, copy, log etc. In the case we are talking about, there is no Neighbor entry object in the first place to associate this UDT.

 Creating a default hostif trap allows the NOS to have the same workflow as they do for existing default traps.

Reply:
Hi Ravi,

I see that there is reference to user define trap in the route entry. Today we use this reference to trap the subnet routes. I am still not clear why this will not work and why need another trap type.
And the comments in SAI_HOSTIF_USER_DEFINED_TRAP_TYPE_NEIGHBOR clearly says that it can be used either for ACL action or for neighbor table miss.

    /**
     * @brief Generate User Defined Trap ID for trap/log actions
     *
     * When it is SAI_NULL_OBJECT_ID, then packet will not be trapped.
     *
     * @type sai_object_id_t
     * @flags CREATE_AND_SET
     * @objects SAI_OBJECT_TYPE_HOSTIF_USER_DEFINED_TRAP
     * @allownull true
     * @default SAI_NULL_OBJECT_ID
     */
    SAI_ROUTE_ENTRY_ATTR_USER_TRAP_ID,

Reply:
Hi @JaiOCP ,

 When an IP address and subnet is assigned to a router interface, the NOS installs a subnet route. The SAI spec says the subnet route will be added with the nexthop as the router interface. In this case, the packet action is NOT trap/log/copy. So, SAI\_ROUTE\_ENTRY\_ATTR\_USER\_TRAP\_ID is NOT valid for a subnet route.

Refer: [https://github.com/opencomputeproject/SAI/blob/d790980b1cdf4eb0213fffd4271d4e502a2591e8/inc/sairoute.h#L77](https://github.com/opencomputeproject/SAI/blob/d790980b1cdf4eb0213fffd4271d4e502a2591e8/inc/sairoute.h#L77)

SONiC adheres to the above SAI definition and uses the RIF as the nexthop for subnet routes. SONiC does not use SAI\_ROUTE\_ENTRY\_ATTR\_USER\_TRAP\_ID for subnet routes.

Reply:
Hi @rck-innovium ,

Lets discuss this in meeting. There is a cleaner way to handle this today and I feel we do not need a new attribute.

Here is a workflow that is per the spec.
router entry for subnet route comes with a RIF for which two things need to happen.
1. SAI_ROUTE_ENTRY_ATTR_USER_TRAP_ID: Install the UDT with neighbor miss 
2. SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID: If the NH is the RIF then handle the UDT to vendor specific implementation for traping the subnet route.

This approach provides us ability to install different UDT for differente subnet route specially if they need to be metered differently.

Reply:
@JaiOCP  

SAI\_HOSTIF\_TRAP\_TYPE\_NEIGHBOR\_MISS is analogous to SAI\_HOSTIF\_TRAP\_TYPE\_IP2ME. 

The only difference is that SAI\_HOSTIF\_TRAP\_TYPE\_IP2ME applies to packets hitting the IP2ME routes, and SAI\_HOSTIF\_TRAP\_TYPE\_NEIGHBOR\_MISS applies to packets hitting the subnet route (i.e. no neighbor/ARP entry is present).

Today, a NOS following SAI (say SONiC) installs IP2ME routes with NH as CPU port. They do not set SAI\_ROUTE\_ENTRY\_ATTR\_USER\_TRAP\_ID. So, packets hitting any IP2ME route go to the CPU based on the configuration for the SAI\_HOSTIF\_TRAP\_TYPE\_IP2ME. \[Say Workflow#1\] 

 If the NOS wants each IP2ME route to be processed differently, then they can use the workflow you mentioned above, i.e., use SAI\_ROUTE\_ENTRY\_ATTR\_USER\_TRAP\_ID. \[Say Workflow#2\]

In essence, there are two workflows a NOS can follow. SAI currently supports Workflow#2 for all route based traps, and the SAI support for workflow#1 is available only for IP2ME but not neighbor-miss. This proposal is adding the missing piece for neigbor-miss packets.

Reply:
Hi Ravi,

Here is the work flow that is followed in vendor SONiC. 
Create a UDT for neighbor miss and bind it the neighbor entry SAI_NEIGHBOR_ENTRY_ATTR_USER_TRAP_ID,
As you can see this is per neighbor and not per route entry there are no issues.

I am ok with you create HOSTIF_TRAP for neighbor miss but in SONiC I would prefer if this option is chosen based on capability query.

Essentially we have two methods.
1.   SAI_NEIGHBOR_ENTRY_ATTR_USER_TRAP_ID,
2.  SAI_HOSTIF_TRAP_TYPE_NEIGHBOR_MISS

and NOS can query what a given SAI adapter is supporting.